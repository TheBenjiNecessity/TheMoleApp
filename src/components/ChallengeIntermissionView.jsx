import React from 'react';
import TitlePanel from '../common/TitlePanel';

const ChallengeIntermissionView = ({ previousChallenge }) => {
	const { intermissionText } = previousChallenge;
	return (
		<TitlePanel titleText="Challenge Intermission">
			<div>{intermissionText ? <p>{intermissionText}</p> : <p>TODO: generic text</p>}</div>
		</TitlePanel>
	);
};

export default ChallengeIntermissionView;
