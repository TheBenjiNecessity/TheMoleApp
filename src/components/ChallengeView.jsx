import React from 'react';

const ChallengeView = ({ challenge }) => {
	return <div />;
};

export default ChallengeView;
