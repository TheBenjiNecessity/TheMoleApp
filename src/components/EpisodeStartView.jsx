import React from 'react';
import NextPanel from '../common/NextPanel';

const EpisodeStartView = ({ previousEpisode, currentEpisode, onNextCallback }) => {
	return (
		<NextPanel titleText="Welcome" nextButtonText="Start Game" onNext={onNextCallback}>
			{previousEpisode ? (
				<div>TODO: text describing what happened in the previous episode</div>
			) : (
				<div>TODO: describe the first episode</div>
			)}
		</NextPanel>
	);
};

export default EpisodeStartView;
