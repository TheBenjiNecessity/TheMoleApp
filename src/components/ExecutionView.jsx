import React, { useCallback, useEffect, useState } from 'react';
import * as _ from 'lodash';
import styles from './ExecutionView.module.scss';
import clsx from 'clsx';

const EXECUTION_STATE = {
	START: 0, // Before showing the player input (Showing a message about seeing red/green screen)
	PRE_SUBMIT_WAITING: 1, // Player input is showing but typing hasn't started
	SUBMITTING: 2, // Characters are being typed into the input
	POST_SUBMIT_WAITING: 3, // Typing has finished but input hasn't been submitted yet
	LOADING_RESULTS: 4, // Typing has finished and the input has been submited hiding the input (waiting for result)
	SHOWING_RESULTS: 5 // The Red/Green results screen is showing
};

const CHAR_INPUT_TIMES = [ 500, 750, 1000, 1500 ];

const ExecutionView = ({ shuffledPlayers, eliminatedPlayer }) => {
	const [ currentShuffledPlayers, setCurrentShuffledPlayers ] = useState(shuffledPlayers);
	const [ currentInputPlayer, setCurrentInputPlayer ] = useState(shuffledPlayers[0]);
	const [ currentCharacterIndex, setCurrentCharacterIndex ] = useState(-1);
	const [ currentTick, setCurrentTick ] = useState(0);
	const [ currentState, setCurrentState ] = useState(EXECUTION_STATE.START);
	const [ inputValue, setInputValue ] = useState('');

	const showStartParagraph = currentState === EXECUTION_STATE.START;
	const showInput = _.includes(
		[ EXECUTION_STATE.PRE_SUBMIT_WAITING, EXECUTION_STATE.SUBMITTING, EXECUTION_STATE.POST_SUBMIT_WAITING ],
		currentState
	);

	const setNextPlayer = useCallback(
		() => {
			const newShuffledPlayers = currentShuffledPlayers.slice(1);

			setCurrentState(EXECUTION_STATE.PRE_SUBMIT_WAITING);
			setCurrentInputPlayer(newShuffledPlayers[0]);
			setCurrentShuffledPlayers(newShuffledPlayers);
			setCurrentTick(15);
			setCurrentCharacterIndex(0);
			setInputValue('');
		},
		[ currentShuffledPlayers ]
	);

	useEffect(
		() => {
			const timer = setTimeout(() => {
				if (currentTick < 15) {
					setCurrentState(EXECUTION_STATE.START);
				} else if (currentTick >= 15 && currentTick < 20) {
					setCurrentState(EXECUTION_STATE.PRE_SUBMIT_WAITING);
				} else if (currentTick >= 20 && currentTick < 30) {
					setCurrentState(EXECUTION_STATE.SUBMITTING);
				} else if (currentTick >= 30 && currentTick < 35) {
					setCurrentState(EXECUTION_STATE.POST_SUBMIT_WAITING);
				} else if (currentTick >= 35 && currentTick < 40) {
					setCurrentState(EXECUTION_STATE.LOADING_RESULTS);
				} else if (currentTick >= 40 && currentTick < 45) {
					setCurrentState(EXECUTION_STATE.SHOWING_RESULTS);
				} else if (currentTick >= 45) {
					setNextPlayer();
				}

				if (currentState !== EXECUTION_STATE.SUBMITTING && currentTick < 45) {
					setCurrentTick(currentTick + 1);
				}
			}, 1000);
			return () => clearTimeout(timer);
		},
		[ currentState, currentTick, setNextPlayer ]
	);

	useEffect(
		() => {
			if (currentState === EXECUTION_STATE.SUBMITTING) {
				if (currentCharacterIndex >= currentInputPlayer.name.length) {
					setCurrentTick(30);
				} else {
					const randomInterval = CHAR_INPUT_TIMES[_.random(CHAR_INPUT_TIMES.length)];
					const timer = setTimeout(() => {
						setInputValue(currentInputPlayer.name.slice(0, currentCharacterIndex + 1));
						setCurrentCharacterIndex(currentCharacterIndex + 1);
					}, randomInterval);
					return () => clearTimeout(timer);
				}
			}
		},
		[ currentCharacterIndex, currentInputPlayer.name, currentState, setNextPlayer ]
	);

	return (
		<div
			className={clsx([
				styles.execution,
				currentState === EXECUTION_STATE.SHOWING_RESULTS && styles['execution-end'],
				eliminatedPlayer.name === currentInputPlayer.name && styles.fail,
				eliminatedPlayer.name !== currentInputPlayer.name && styles.success
			])}
		>
			{currentTick}
			{showStartParagraph && (
				<p>
					In a moment, each player's name will be input into the input box one at a time. If you see a green
					screen after your name has been entered you are safe for this episode. If, however, you see a red
					screen after your name has been entered then you have been eliminated from the game.
				</p>
			)}
			{showInput && <input className={styles.input} type="text" value={inputValue} />}
		</div>
	);
};

export default ExecutionView;

// What if each player can see a green screen on their device if they have already been cleared
