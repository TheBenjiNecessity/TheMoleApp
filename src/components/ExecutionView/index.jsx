import React, { useCallback, useEffect, useMemo, useState } from 'react';
import * as _ from 'lodash';
import styles from './styles.module.scss';
import clsx from 'clsx';
import TextInput from '../../common/TextInput';
import Logo from '../../common/Logo';
import roomSocketService from '../../services/socket-services/room-socket.service';

const EXECUTION_STATE = {
	START: 0, // Before showing the player input (Showing a message about seeing red/green screen)
	PRE_SUBMIT_WAITING: 1, // Player input is showing but typing hasn't started
	SUBMITTING: 2, // Characters are being typed into the input
	POST_SUBMIT_WAITING: 3, // Typing has finished but input hasn't been submitted yet
	LOADING_RESULTS: 4, // Typing has finished and the input has been submited hiding the input (waiting for result)
	SHOWING_RESULTS: 5 // The Red/Green results screen is showing
};

const CHAR_INPUT_TIMES = [ 900, 750, 1000, 800 ];

const ExecutionView = ({ roomcode, shuffledPlayers, eliminatedPlayer }) => {
	const allBeforeEl = useMemo(
		() => {
			return _.slice(
				shuffledPlayers,
				0,
				_.findIndex(shuffledPlayers, (player) => player.name === eliminatedPlayer.name) + 1
			);
		},
		[ eliminatedPlayer, shuffledPlayers ]
	);

	const [ currentShuffledPlayers, setCurrentShuffledPlayers ] = useState(allBeforeEl);
	const [ currentInputPlayer, setCurrentInputPlayer ] = useState(shuffledPlayers[0]);
	const [ currentCharacterIndex, setCurrentCharacterIndex ] = useState(-1);
	const [ currentTick, setCurrentTick ] = useState(0);
	const [ currentState, setCurrentState ] = useState(EXECUTION_STATE.START);
	const [ inputValue, setInputValue ] = useState('');
	const [ cursor, setCursor ] = useState('|');

	const showStartParagraph = currentState === EXECUTION_STATE.START;
	const showFullLogo = currentState === EXECUTION_STATE.SHOWING_RESULTS;
	const showCornerLogo = currentState !== EXECUTION_STATE.SHOWING_RESULTS;
	const showInput = _.includes(
		[ EXECUTION_STATE.PRE_SUBMIT_WAITING, EXECUTION_STATE.SUBMITTING, EXECUTION_STATE.POST_SUBMIT_WAITING ],
		currentState
	);

	const setNextPlayer = useCallback(
		() => {
			if (currentShuffledPlayers.length >= 2) {
				const newShuffledPlayers = currentShuffledPlayers.slice(1);

				setCurrentState(EXECUTION_STATE.PRE_SUBMIT_WAITING);
				setCurrentInputPlayer(newShuffledPlayers[0]);
				setCurrentShuffledPlayers(newShuffledPlayers);
				setCurrentTick(15);
				setCurrentCharacterIndex(0);
				setInputValue('');
			} else {
				roomSocketService.moveNext(roomcode);
			}
		},
		[ currentShuffledPlayers, roomcode ]
	);

	useEffect(
		() => {
			const timer = setTimeout(() => {
				if (currentTick < 15) {
					setCurrentState(EXECUTION_STATE.START);
				} else if (currentTick >= 15 && currentTick < 20) {
					setCurrentState(EXECUTION_STATE.PRE_SUBMIT_WAITING);
				} else if (currentTick >= 20 && currentTick < 30) {
					setCurrentState(EXECUTION_STATE.SUBMITTING);
				} else if (currentTick >= 30 && currentTick < 35) {
					setCurrentState(EXECUTION_STATE.POST_SUBMIT_WAITING);
				} else if (currentTick >= 35 && currentTick < 40) {
					setCurrentState(EXECUTION_STATE.LOADING_RESULTS);
				} else if (currentTick >= 40 && currentTick < 45) {
					setCurrentState(EXECUTION_STATE.SHOWING_RESULTS);
				} else if (currentTick >= 45) {
					setNextPlayer();
				}

				if (currentState !== EXECUTION_STATE.SUBMITTING && currentTick < 45) {
					setCurrentTick(currentTick + 1);
				}
			}, 1000);
			return () => clearTimeout(timer);
		},
		[ currentState, currentTick, setNextPlayer ]
	);

	useEffect(
		() => {
			if (currentState === EXECUTION_STATE.SUBMITTING) {
				if (currentCharacterIndex >= currentInputPlayer.name.length) {
					setCurrentTick(30);
				} else {
					const randomInterval = CHAR_INPUT_TIMES[_.random(CHAR_INPUT_TIMES.length)];
					const timer = setTimeout(() => {
						setInputValue(currentInputPlayer.name.slice(0, currentCharacterIndex + 1));
						setCurrentCharacterIndex(currentCharacterIndex + 1);
					}, randomInterval);
					return () => clearTimeout(timer);
				}
			}
		},
		[ currentCharacterIndex, currentInputPlayer.name, currentState, setNextPlayer ]
	);

	useEffect(
		() => {
			const timer = setTimeout(() => {
				setCursor(cursor === '|' ? '' : '|');
			}, 500);
			return () => clearTimeout(timer);
		},
		[ cursor ]
	);

	return (
		<div
			className={clsx([
				styles.execution,
				currentState === EXECUTION_STATE.SHOWING_RESULTS && styles['execution-end'],
				eliminatedPlayer.name === currentInputPlayer.name && styles.fail,
				eliminatedPlayer.name !== currentInputPlayer.name && styles.success
			])}
		>
			<div className="panel abs-centered-panel hv-centered-panel">
				{showStartParagraph && (
					<div className={styles['start-paragraph']}>
						<p>
							In a moment, each player's name will be input into the input box one at a time. If you see a
							green screen after your name has been entered you are safe for this episode. If, however,
							you see a red screen after your name has been entered then you have been eliminated from the
							game.
						</p>
					</div>
				)}
				{showInput && (
					<label className={styles.inputLabel}>
						<span className={styles.text}>Name:</span>
						<span className={styles.input}>
							<TextInput value={`${inputValue}${cursor}`} />
						</span>
					</label>
				)}
				{showFullLogo && (
					<div
						className={clsx([
							eliminatedPlayer.name === currentInputPlayer.name && styles.fail,
							eliminatedPlayer.name !== currentInputPlayer.name && styles.success
						])}
					>
						<Logo size={400} logoColor="white" />
					</div>
				)}
			</div>
			{showCornerLogo && (
				<div className={clsx([ styles.logo, styles['corner-logo'] ])}>
					<Logo size={200} textSize={45} />
				</div>
			)}
		</div>
	);
};

export default ExecutionView;

// What if each player can see a green screen on their device if they have already been cleared
