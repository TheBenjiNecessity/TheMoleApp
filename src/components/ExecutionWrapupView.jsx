import React from 'react';
import NextPanel from '../common/NextPanel';

const ExecutionWrapupView = ({ eliminatedPlayer, onNext }) => {
	return (
		<NextPanel titleText="Welcome" nextButtonText="Start Game" onNext={onNext}>
			<div>TODO: text describing the game {eliminatedPlayer.name}</div>
		</NextPanel>
	);
};

export default ExecutionWrapupView;
