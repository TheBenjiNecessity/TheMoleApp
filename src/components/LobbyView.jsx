import React from 'react';
//import { useTranslation } from 'react-i18next';
import AgreePanel from '../common/AgreePanel';
import NextPanel from '../common/NextPanel';

import PlayerListView from '../common/PlayerListView';

const LobbyView = ({ room, onNext }) => {
	return (
		<NextPanel titleText="Lobby" nextButtonText="Start" onNext={onNext}>
			<div className="form-group pl-xs-0 pr-xs-0 mt-0 col-sm-6">
				<label>Room Code:</label>
				<div className="room-code">{room ? room.roomcode : 'No Code'}</div>
			</div>
			<div className="form-group pl-xs-0 pr-xs-0 mt-0 col-sm-6">
				<label>Players:</label>
				<div className="player-list">
					{room && room.players && room.players.length ? (
						room.players.map((player) => <PlayerListView key={player.name} player={player} />)
					) : (
						<div />
					)}
				</div>
			</div>
			<AgreePanel room={room} />
		</NextPanel>
	);
};

export default LobbyView;
