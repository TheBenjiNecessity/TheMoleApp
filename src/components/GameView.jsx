import React from 'react';
import WelcomeView from './steps/WelcomeView';
import EpisodeStartView from './steps/EpisodeStartView';
import ChallengeView from './steps/ChallengeView';
import ChallengeIntermissionView from './steps/ChallengeIntermissionView';
import PreQuizIntermissionView from './steps/PreQuizIntermissionView';
import QuizView from './steps/QuizView';
import PostQuizIntermissionView from './steps/PostQuizIntermissionView';
import ExecutionView from './steps/ExecutionView';
import ExecutionWrapupView from './steps/ExecutionWrapupView';
import Room from '../models/room.model';
import LobbyView from './steps/LobbyView';
import MoleRevealView from './steps/MoleRevealView';

const GameView = ({ room }) => {
	switch (room.state) {
		case Room.ROOM_STATES.LOBBY:
			return <LobbyView room={room} />;
		case Room.ROOM_STATES.WELCOME:
			return <WelcomeView room={room} />;
		case Room.ROOM_STATES.MOLE_REVEAL:
			return <MoleRevealView />;
		case Room.ROOM_STATES.EPISODE_START:
			return <EpisodeStartView room={room} />;
		case Room.ROOM_STATES.IN_CHALLENGE:
			return <ChallengeView room={room} />;
		case Room.ROOM_STATES.CHALLENGE_INTERMISSION:
			return <ChallengeIntermissionView room={room} />;
		case Room.ROOM_STATES.PRE_QUIZ_INTERMISSION:
			return <PreQuizIntermissionView room={room} />;
		case Room.ROOM_STATES.IN_QUIZ:
			return <QuizView room={room} />;
		case Room.ROOM_STATES.POST_QUIZ_INTERMISSION:
			return <PostQuizIntermissionView room={room} />;
		case Room.ROOM_STATES.EXECUTION:
			return <ExecutionView room={room} />;
		case Room.ROOM_STATES.EXECUTION_WRAPUP:
			return <ExecutionWrapupView room={room} />;
		default:
			return null;
	}
};

export default GameView;
