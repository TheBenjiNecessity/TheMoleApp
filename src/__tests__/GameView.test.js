import React from 'react';
import { mount, shallow } from 'enzyme';
import GameView from '../components/GameView';
import '../locales/i18n';
import LobbyView from '../components/steps/LobbyView';
import roomSample from './samples/room.sample';
import WelcomeView from '../components/steps/WelcomeView';
import MoleRevealView from '../components/steps/MoleRevealView';
import EpisodeStartView from '../components/steps/EpisodeStartView';
import ChallengeView from '../components/steps/ChallengeView';
import ChallengeIntermissionView from '../components/steps/ChallengeIntermissionView';
import PreQuizIntermissionView from '../components/steps/PreQuizIntermissionView';
import QuizView from '../components/steps/QuizView';
import PostQuizIntermissionView from '../components/steps/PostQuizIntermissionView';
import ExecutionView from '../components/steps/ExecutionView';
import ExecutionWrapupView from '../components/steps/ExecutionWrapupView';

describe('MyComponent', () => {
	let room = {};

	beforeEach(() => {
		room = roomSample.sampleRoom();
	});

	it('should render correctly', () => {
		const component = shallow(<GameView room={room} />);
		expect(component).toMatchSnapshot();
	});

	it('should render lobby', () => {
		room.state = 'lobby';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(LobbyView);
	});

	it('should render welcome', () => {
		room.state = 'game-welcome';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(WelcomeView);
	});

	it('should render mole reveal', () => {
		room.state = 'mole-reveal';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(MoleRevealView);
	});

	it('should render episode start', () => {
		room.state = 'episode-start';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(EpisodeStartView);
	});

	it('should render challenge', () => {
		room.state = 'in-challenge';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(ChallengeView);
	});

	it('should render challenge intermission', () => {
		room.state = 'challenge-intermission';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(ChallengeIntermissionView);
	});

	it('should render pre quiz', () => {
		room.state = 'pre-quiz-intermission';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(PreQuizIntermissionView);
	});

	it('should render quiz', () => {
		room.state = 'in-quiz';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(QuizView);
	});

	it('should render post quiz', () => {
		room.state = 'post-quiz-intermission';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(PostQuizIntermissionView);
	});

	it('should render execution', () => {
		room.state = 'execution';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(ExecutionView);
	});

	it('should render execution-wrapup', () => {
		room.state = 'execution-wrapup';
		const component = shallow(<GameView room={room} />);
		expect(component.type()).toEqual(ExecutionWrapupView);
	});
});
