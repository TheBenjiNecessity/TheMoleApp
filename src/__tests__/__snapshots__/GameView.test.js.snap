// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent should render correctly 1`] = `
<LobbyView
  data-testid="lobby"
  room={
    Room {
      "agreedPlayers": Array [],
      "episodes": Array [],
      "players": Array [
        Object {
          "name": "test1",
        },
        Object {
          "name": "test2",
        },
        Object {
          "name": "test3",
        },
        Object {
          "name": "test4",
        },
        Object {
          "name": "test5",
        },
        Object {
          "name": "test6",
        },
        Object {
          "name": "test7",
        },
        Object {
          "name": "test8",
        },
        Object {
          "name": "test9",
        },
        Object {
          "name": "test10",
        },
      ],
      "raisedHands": Object {},
      "roomcode": "TEST",
      "state": "lobby",
    }
  }
/>
`;
